bindings/python/classad3/Expression.py

```diff
@@ -1 +1,106 @@
+from typing import List
 
+from .classad3_impl import _canonicalize_expr_string
+from .classad3_impl import _evaluate
+from .classad3_impl import _flatten
+from .classad3_impl import _external_refs
+from .classad3_impl import _internal_refs
+
+import classad3
+from classad3 import ClassAd
+
+
+class Expression:
+    '''
+    Expression class docstring.
+    '''
+
+
+    def __init__(self, expr_string):
+        '''
+        Constructor docstrings are ignored?
+        '''
+        self.expr_string = _canonicalize_expr_string(expr_string)
+
+
+    def asNewClassAd(self) -> str:
+        '''
+        Serialize this expression into the "new" ClassAd format.
+
+        :return:  This expression serialized in the "new" ClassAd format.
+        '''
+        return self.expr_string
+
+
+    # Note that since the simplifid Expression is just an opaque handle
+    # -- in fact, the canonical string representation of the expression --
+    # we don't have to worry about object lifetimes here.
+    def evaluate(self, **kwargs):
+        '''
+        Evaluate this expression and return the corresponding ClassAd
+        value, or raise :class:`ClassAdValueError` if the expression's
+        value is ``error``.
+
+        :param ClassAd my: Optionally, the ``MY`` :class:`ClassAd`.
+        :param ClassAd target: Optionally, the ``TARGET`` :class:`ClassAd`.
+        :return: The ClassAd value result of evaluating this expression.
+        '''
+
+        my = kwargs.get("my")
+        target = kwargs.get("target")
+
+        my_string = "" if my is None else my.asNewClassAd()
+        target_string = "" if target is None else target.asNewClassAd()
+
+        answer_blob = _evaluate(self.expr_string, my_string, target_string)
+        object_blob = '{ "rv": ' + answer_blob + '}'
+        c = classad3.ClassAd._fromJSON(object_blob)
+        return c['rv']
```

Comment:
Lots of questions...

1. If you remove the leading `classad3`...
```suggestion
        answer_blob = _evaluate(self.expr_string, my_string, target_string)
        object_blob = '{ "rv": ' + answer_blob + '}'
        c = ClassAd._fromJSON(object_blob)
        return c['rv']
```
most importantly, does it work, and less importantly, can you get rid of `import classad3`?

2. What are the two middle lines here trying to accomplish? Making sure the answer is a valid ClassAd value type? Seems weird to have to go through the JSON parser here.

3. Sidenote, and maybe this is why you did the imports this way, did you run into any problems with circular imports?

Reply:
[I'll answer 1 and 3 later.]

`answer_blob` is actually a blob, that is, a JSON string, and needs to be converted into an `Expression`; this is the easiest (albeit a clumsy) way to do that.  Part of the `classad3` big idea is that ClassAds and expressions only get passed in and out of the C++ library as strings: no pointers to complicated data structures with unfortunate ownership models / lifetime policies.

Reply:
1. I can remove the `from classad3 import ClassAd`, but I can't change anything else.
3. I didn't run into any problems with circular imports, probably because of import caching.

Reply:
3 - Although note that at least for now, the type hints are string literals and not type references.